File: packages/typescript/src/cloudlink.ts
import {
  CONTRACT_FAILURE_CODES,
  type ContractFailureCode,
} from "./failure.js";

const maximumUint64 = 18_446_744_073_709_551_615n;
const digestPattern = /^sha256:[0-9a-f]{64}$/u;
const semverPattern =
  /^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)(?:-(?:0|[1-9][0-9]*|[0-9]*[A-Za-z-][0-9A-Za-z-]*)(?:\.(?:0|[1-9][0-9]*|[0-9]*[A-Za-z-][0-9A-Za-z-]*))*)?(?:\+[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?$/u;

export interface CloudLinkSessionContext {
  readonly credentialGeneration: string;
  readonly gatewayId: string;
  readonly sessionEpoch: string;
  readonly sessionId: string;
}

export interface CloudLinkFixtureContext {
  readonly currentSession?: CloudLinkSessionContext;
  readonly priorAcceptedDelivery?: string;
}

export type CloudLinkFixtureResult =
  | { readonly accepted: true; readonly failureCode?: never }
  | { readonly accepted: false; readonly failureCode: ContractFailureCode };

interface JsonRecord {
  readonly [key: string]: unknown;
  readonly aether_version?: unknown;
  readonly algorithm?: unknown;
  readonly batch_id?: unknown;
  readonly client_nonce?: unknown;
  readonly cloud_nonce?: unknown;
  readonly cloud_signature?: unknown;
  readonly credential_binding?: unknown;
  readonly credential_generation?: unknown;
  readonly cursors?: unknown;
  readonly delivery?: unknown;
  readonly digest?: unknown;
  readonly expires_at_ms?: unknown;
  readonly gateway_id?: unknown;
  readonly gateway_signature?: unknown;
  readonly key_id?: unknown;
  readonly manifest?: unknown;
  readonly message_kind?: unknown;
  readonly origin_model?: unknown;
  readonly payload?: unknown;
  readonly position?: unknown;
  readonly protocol_version?: unknown;
  readonly resume?: unknown;
  readonly session_epoch?: unknown;
  readonly session_id?: unknown;
  readonly signature?: unknown;
  readonly stream_epoch?: unknown;
  readonly stream_id?: unknown;
}

function rejected(failureCode: ContractFailureCode): CloudLinkFixtureResult {
  return { accepted: false, failureCode };
}

function isRecord(value: unknown): value is JsonRecord {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function isValidationResult(value: JsonRecord | CloudLinkFixtureResult): value is CloudLinkFixtureResult {
  return typeof value.accepted === "boolean";
}

function parseRecord(raw: string): JsonRecord | CloudLinkFixtureResult {
  let value: unknown;
  try {
    value = JSON.parse(raw) as unknown;
  } catch {
    return rejected(CONTRACT_FAILURE_CODES.JSON_SYNTAX_ERROR);
  }
  return isRecord(value)
    ? value
    : rejected(CONTRACT_FAILURE_CODES.JSON_SYNTAX_ERROR);
}

function stringField(record: JsonRecord, key: string): string | undefined {
  const value = record[key];
  return typeof value === "string" ? value : undefined;
}

function unknownField(
  record: JsonRecord,
  allowed: ReadonlySet<string>,
): CloudLinkFixtureResult | undefined {
  return Object.keys(record).some((key) => !allowed.has(key))
    ? rejected(CONTRACT_FAILURE_CODES.UNKNOWN_FIELD)
    : undefined;
}

function uint64Failure(value: unknown): ContractFailureCode | undefined {
  if (typeof value !== "string") {
    return CONTRACT_FAILURE_CODES.INTEGER_NON_CANONICAL;
  }
  if (new TextEncoder().encode(value).byteLength > 20) {
    return CONTRACT_FAILURE_CODES.INTEGER_OUT_OF_RANGE;
  }
  if (!/^(0|[1-9][0-9]*)$/u.test(value)) {
    return CONTRACT_FAILURE_CODES.INTEGER_NON_CANONICAL;
  }
  return BigInt(value) > maximumUint64
    ? CONTRACT_FAILURE_CODES.INTEGER_OUT_OF_RANGE
    : undefined;
}

function firstUint64Failure(
  record: JsonRecord,
  fields: readonly string[],
): ContractFailureCode | undefined {
  for (const field of fields) {
    const failure = uint64Failure(record[field]);
    if (failure !== undefined) {
      return failure;
    }
  }
  return undefined;
}

function signatureIsValid(value: unknown): boolean {
  if (!isRecord(value)) {
    return false;
  }
  return (
    unknownField(
      value,
      new Set(["algorithm", "key_id", "signature"]),
    ) === undefined &&
    value.algorithm === "Ed25519" &&
    typeof value.key_id === "string" &&
    /^[A-Za-z0-9_-]{86}$/u.test(String(value.signature))
  );
}

function cursorsConflict(value: unknown): boolean {
  if (!Array.isArray(value)) {
    return false;
  }
  const identities = new Set<string>();
  for (const cursor of value) {
    if (!isRecord(cursor)) {
      continue;
    }
    const identity = JSON.stringify([cursor.stream_id, cursor.stream_epoch]);
    if (identities.has(identity)) {
      return true;
    }
    identities.add(identity);
  }
  return false;
}

function staleSession(
  message: JsonRecord,
  current: CloudLinkSessionContext | undefined,
): boolean {
  return (
    current !== undefined &&
    (message.gateway_id !== current.gatewayId ||
      message.session_id !== current.sessionId ||
      message.session_epoch !== current.sessionEpoch ||
      message.credential_generation !== current.credentialGeneration)
  );
}

function replayIdentity(message: JsonRecord): readonly unknown[] | undefined {
  const delivery = message.delivery;
  return isRecord(delivery)
    ? [
        message.gateway_id,
        delivery.stream_id,
        delivery.stream_epoch,
        delivery.position,
      ]
    : undefined;
}

function arraysEqual(left: readonly unknown[], right: readonly unknown[]): boolean {
  return left.length === right.length && left.every((value, index) => value === right[index]);
}

function validateDeliveryContext(
  message: JsonRecord,
  context: CloudLinkFixtureContext,
): CloudLinkFixtureResult | undefined {
  if (context.priorAcceptedDelivery === undefined) {
    return undefined;
  }
  const priorResult = parseRecord(context.priorAcceptedDelivery);
  if (!isValidationResult(priorResult)) {
    const candidateIdentity = replayIdentity(message);
    const priorIdentity = replayIdentity(priorResult);
    if (
      candidateIdentity !== undefined &&
      priorIdentity !== undefined &&
      arraysEqual(candidateIdentity, priorIdentity)
    ) {
      const delivery = message.delivery;
      const priorDelivery = priorResult.delivery;
      if (isRecord(delivery) && isRecord(priorDelivery)) {
        const payloadMatches =
          JSON.stringify(message.payload) === JSON.stringify(priorResult.payload);
        if (
          (delivery.digest === priorDelivery.digest && !payloadMatches) ||
          (delivery.digest !== priorDelivery.digest && payloadMatches)
        ) {
          return rejected(CONTRACT_FAILURE_CODES.DIGEST_MISMATCH);
        }
        if (
          delivery.digest !== priorDelivery.digest ||
          delivery.batch_id !== priorDelivery.batch_id
        ) {
          return rejected(CONTRACT_FAILURE_CODES.DIGEST_CONFLICT);
        }
      }
    }
  }
  return undefined;
}

function validateSessionHello(message: JsonRecord): CloudLinkFixtureResult {
  const unknown = unknownField(
    message,
    new Set([
      "challenge_id",
      "client_nonce",
      "credential_binding",
      "gateway_id",
      "gateway_key_id",
      "gateway_signature",
      "message_kind",
      "offered_protocol_versions",
      "protocol",
      "resume",
      "schema",
    ]),
  );
  if (unknown !== undefined) {
    return unknown;
  }
  if (!/^[A-Za-z0-9_-]{43}$/u.test(String(message.client_nonce))) {
    return rejected(CONTRACT_FAILURE_CODES.FIELD_BOUND);
  }
  const binding = message.credential_binding;
  if (!isRecord(binding)) {
    return rejected(CONTRACT_FAILURE_CODES.UNKNOWN_FIELD);
  }
  if (binding.origin_model === "gateway-signed") {
    if (message.gateway_signature === undefined) {
      return rejected(CONTRACT_FAILURE_CODES.AUTHENTICATION_REQUIRED);
    }
    if (!signatureIsValid(message.gateway_signature)) {
      return rejected(CONTRACT_FAILURE_CODES.AUTHENTICATION_INVALID);
    }
  } else if (binding.origin_model !== "trusted-connector-broker-attestation") {
    return rejected(CONTRACT_FAILURE_CODES.UNKNOWN_FIELD);
  }
  return cursorsConflict(message.resume)
    ? rejected(CONTRACT_FAILURE_CODES.CURSOR_CONFLICT)
    : { accepted: true };
}

function validateSessionChallengeRequest(message: JsonRecord): CloudLinkFixtureResult {
  const unknown = unknownField(
    message,
    new Set([
      "client_nonce",
      "credential_binding",
      "gateway_id",
      "message_kind",
      "offered_protocol_versions",
      "protocol",
      "resume",
      "schema",
    ]),
  );
  if (unknown !== undefined) {
    return unknown;
  }
  if (!/^[A-Za-z0-9_-]{43}$/u.test(String(message.client_nonce))) {
    return rejected(CONTRACT_FAILURE_CODES.FIELD_BOUND);
  }
  const binding = message.credential_binding;
  if (!isRecord(binding)) {
    return rejected(CONTRACT_FAILURE_CODES.UNKNOWN_FIELD);
  }
  const bindingUnknown = unknownField(
    binding,
    new Set(["credential_id", "generation"]),
  );
  if (bindingUnknown !== undefined) {
    return bindingUnknown;
  }
  const generationFailure = uint64Failure(binding["generation"]);
  if (generationFailure !== undefined) {
    return rejected(generationFailure);
  }
  return cursorsConflict(message.resume)
    ? rejected(CONTRACT_FAILURE_CODES.CURSOR_CONFLICT)
    : { accepted: true };
}

function validateSessionChallenge(message: JsonRecord): CloudLinkFixtureResult {
  const unknown = unknownField(
    message,
    new Set([
      "challenge_id",
      "cloud_nonce",
      "cloud_signature",
      "expires_at_ms",
      "gateway_id",
      "issued_at_ms",
      "message_kind",
      "protocol",
      "schema",
    ]),
  );
  if (unknown !== undefined) {
    return unknown;
  }
  const failure = firstUint64Failure(message, ["issued_at_ms", "expires_at_ms"]);
  if (failure !== undefined) {
    return rejected(failure);
  }
  return /^[A-Za-z0-9_-]{43}$/u.test(String(message.cloud_nonce)) &&
    signatureIsValid(message.cloud_signature)
    ? { accepted: true }
    : rejected(CONTRACT_FAILURE_CODES.FIELD_BOUND);
}

function validateHeartbeat(
  message: JsonRecord,
  context: CloudLinkFixtureContext,
): CloudLinkFixtureResult {
  const unknown = unknownField(
    message,
    new Set([
      "credential_generation",
      "cursors",
      "gateway_id",
      "message_authentication",
      "message_kind",
      "observed_at_ms",
      "protocol",
      "protocol_version",
      "schema",
      "session_epoch",
      "session_id",
    ]),
  );
  if (unknown !== undefined) {
    return unknown;
  }
  if (message.protocol_version !== "1.0") {
    return rejected(CONTRACT_FAILURE_CODES.UNSUPPORTED_VERSION);
  }
  const failure = firstUint64Failure(message, [
    "session_epoch",
    "credential_generation",
    "observed_at_ms",
  ]);
  if (failure !== undefined) {
    return rejected(failure);
  }
  if (Array.isArray(message.cursors)) {
    for (const cursor of message.cursors) {
      if (isRecord(cursor)) {
        const cursorFailure = firstUint64Failure(cursor, [
          "stream_epoch",
          "acknowledged_position",
        ]);
        if (cursorFailure !== undefined) {
          return rejected(cursorFailure);
        }
      }
    }
  }
  if (staleSession(message, context.currentSession)) {
    return rejected(CONTRACT_FAILURE_CODES.STALE_SESSION);
  }
  return cursorsConflict(message.cursors)
    ? rejected(CONTRACT_FAILURE_CODES.CURSOR_CONFLICT)
    : { accepted: true };
}

function validateDurableAck(
  message: JsonRecord,
  context: CloudLinkFixtureContext,
): CloudLinkFixtureResult {
  if (!digestPattern.test(String(message.digest))) {
    return rejected(CONTRACT_FAILURE_CODES.INVALID_DIGEST);
  }
  const failure = firstUint64Failure(message, [
    "session_epoch",
    "credential_generation",
    "stream_epoch",
    "acknowledged_position",
    "acknowledged_at_ms",
  ]);
  if (failure !== undefined) {
    return rejected(failure);
  }
  return staleSession(message, context.currentSession)
    ? rejected(CONTRACT_FAILURE_CODES.STALE_SESSION)
    : { accepted: true };
}

function validateSessionAccepted(message: JsonRecord): CloudLinkFixtureResult {
  const failure = firstUint64Failure(message, [
    "session_epoch",
    "credential_generation",
    "server_time_ms",
    "heartbeat_interval_ms",
  ]);
  if (failure !== undefined) {
    return rejected(failure);
  }
  return cursorsConflict(message.resume)
    ? rejected(CONTRACT_FAILURE_CODES.CURSOR_CONFLICT)
    : { accepted: true };
}

function validateEnvelope(
  message: JsonRecord,
  context: CloudLinkFixtureContext,
): CloudLinkFixtureResult {
  const unknown = unknownField(
    message,
    new Set([
      "credential_generation",
      "delivery",
      "expires_at_ms",
      "gateway_id",
      "message_authentication",
      "message_kind",
      "payload",
      "protocol",
      "protocol_version",
      "schema",
      "sent_at_ms",
      "session_epoch",
      "session_id",
      "traceparent",
    ]),
  );
  if (unknown !== undefined) {
    return unknown;
  }
  if (message.protocol_version !== "1.0") {
    return rejected(CONTRACT_FAILURE_CODES.UNSUPPORTED_VERSION);
  }
  const failure = firstUint64Failure(message, [
    "session_epoch",
    "credential_generation",
    "sent_at_ms",
    ...(message.expires_at_ms === undefined ? [] : ["expires_at_ms"]),
  ]);
  if (failure !== undefined) {
    return rejected(failure);
  }
  const delivery = message.delivery;
  if (!isRecord(delivery)) {
    return rejected(CONTRACT_FAILURE_CODES.UNKNOWN_FIELD);
  }
  const deliveryFailure = firstUint64Failure(delivery, ["stream_epoch", "position"]);
  if (deliveryFailure !== undefined) {
    return rejected(deliveryFailure);
  }
  if (!digestPattern.test(String(delivery.digest))) {
    return rejected(CONTRACT_FAILURE_CODES.INVALID_DIGEST);
  }
  if (message.message_kind === "runtime-manifest-report") {
    const payload = message.payload;
    const manifest = isRecord(payload) ? payload.manifest : undefined;
    if (
      !isRecord(manifest) ||
      !semverPattern.test(String(manifest.aether_version))
    ) {
      return rejected(CONTRACT_FAILURE_CODES.SEMVER_INVALID);
    }
  }
  const contextual = validateDeliveryContext(message, context);
  if (contextual !== undefined) {
    return contextual;
  }
  return staleSession(message, context.currentSession)
    ? rejected(CONTRACT_FAILURE_CODES.STALE_SESSION)
    : { accepted: true };
}

/**
 * Executes the published CloudLink alpha fixture surface.
 *
 * This experimental binding deliberately reports only the stable contractual
 * code. It does not claim production authentication or persistence.
 */
export function validateCloudLinkFixture(
  raw: string,
  context: CloudLinkFixtureContext = {},
): CloudLinkFixtureResult {
  const parsed = parseRecord(raw);
  if (isValidationResult(parsed)) {
    return parsed;
  }
  const kind = stringField(parsed, "message_kind");
  switch (kind) {
    case "session-challenge-request":
      return validateSessionChallengeRequest(parsed);
    case "session-hello":
      return validateSessionHello(parsed);
    case "session-challenge":
      return validateSessionChallenge(parsed);
    case "session-accepted":
      return validateSessionAccepted(parsed);
    case "heartbeat":
    case "heartbeat-ack":
      return validateHeartbeat(parsed, context);
    case "durable-ack":
      return validateDurableAck(parsed, context);
    case "replay-request":
      return { accepted: true };
    case "runtime-manifest-report":
    case "telemetry-batch":
    case "data-loss":
      return validateEnvelope(parsed, context);
    default:
      return rejected(CONTRACT_FAILURE_CODES.UNSUPPORTED_VERSION);
  }
}

File: packages/rust/src/cloudlink.rs
use std::collections::BTreeSet;

use serde_json::{Map, Value, json};
use sha2::{Digest, Sha256};

use crate::{ContractFailure, ContractFailureCode};

const MAXIMUM_UINT64: u128 = u64::MAX as u128;
const DIGEST_PREFIX: &str = "sha256:";

/// Current `CloudLink` session facts supplied to contextual fixture validation.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct CloudLinkSessionContext<'a> {
    pub credential_generation: &'a str,
    pub gateway_id: &'a str,
    pub session_epoch: &'a str,
    pub session_id: &'a str,
}

/// Explicit context for replay and session-sensitive `CloudLink` fixtures.
#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct CloudLinkFixtureContext<'a> {
    pub current_session: Option<CloudLinkSessionContext<'a>>,
    pub prior_accepted_delivery: Option<&'a str>,
}

/// Stable language-neutral outcome for one public fixture.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct CloudLinkValidationResult {
    pub accepted: bool,
    pub failure_code: Option<&'static str>,
}

impl CloudLinkValidationResult {
    /// Mirrors `Result::is_ok` for consumers that only need acceptance.
    #[must_use]
    pub const fn is_ok(self) -> bool {
        self.accepted
    }

    /// Returns the typed stable failure, when validation rejected the input.
    #[must_use]
    pub fn err(self) -> Option<ContractFailure> {
        self.failure_code
            .map(|code| ContractFailure::new(failure_code(code)))
    }
}

/// Accepted input forms for the copy-only fixture context.
pub trait CloudLinkContextInput<'a> {
    fn into_context(self) -> CloudLinkFixtureContext<'a>;
}

impl<'a> CloudLinkContextInput<'a> for CloudLinkFixtureContext<'a> {
    fn into_context(self) -> CloudLinkFixtureContext<'a> {
        self
    }
}

impl<'a> CloudLinkContextInput<'a> for &CloudLinkFixtureContext<'a> {
    fn into_context(self) -> CloudLinkFixtureContext<'a> {
        *self
    }
}

fn failure_code(value: &str) -> ContractFailureCode {
    if value == "UNKNOWN_FIELD" {
        return ContractFailureCode::UnknownField;
    }
    match value {
        "INTEGER_NON_CANONICAL" => ContractFailureCode::IntegerNonCanonical,
        "INTEGER_OUT_OF_RANGE" => ContractFailureCode::IntegerOutOfRange,
        "FIELD_BOUND" => ContractFailureCode::FieldBound,
        "UNSUPPORTED_VERSION" => ContractFailureCode::UnsupportedVersion,
        "INVALID_DIGEST" => ContractFailureCode::InvalidDigest,
        "DIGEST_MISMATCH" => ContractFailureCode::DigestMismatch,
        "DIGEST_CONFLICT" => ContractFailureCode::DigestConflict,
        "STALE_SESSION" => ContractFailureCode::StaleSession,
        "CURSOR_CONFLICT" => ContractFailureCode::CursorConflict,
        "JSON_SYNTAX_ERROR" => ContractFailureCode::JsonSyntaxError,
        "AUTHENTICATION_REQUIRED" => ContractFailureCode::AuthenticationRequired,
        "AUTHENTICATION_INVALID" => ContractFailureCode::AuthenticationInvalid,
        "SEMVER_INVALID" => ContractFailureCode::SemverInvalid,
        "DATA_LOSS_RANGE_INVALID" => ContractFailureCode::DataLossRangeInvalid,
        _ => ContractFailureCode::UnknownField,
    }
}

fn object(value: &Value) -> Option<&Map<String, Value>> {
    value.as_object()
}

fn unknown_field(value: &Map<String, Value>, allowed: &[&str]) -> bool {
    value.keys().any(|key| !allowed.contains(&key.as_str()))
}

fn canonical_u64_failure(value: Option<&Value>) -> Option<&'static str> {
    let Some(value) = value.and_then(Value::as_str) else {
        return Some("INTEGER_NON_CANONICAL");
    };
    if value.len() > 20 {
        return Some("INTEGER_OUT_OF_RANGE");
    }
    if value.is_empty()
        || !value.bytes().all(|byte| byte.is_ascii_digit())
        || (value.len() > 1 && value.starts_with('0'))
    {
        return Some("INTEGER_NON_CANONICAL");
    }
    match value.parse::<u128>() {
        Ok(number) if number <= MAXIMUM_UINT64 => None,
        _ => Some("INTEGER_OUT_OF_RANGE"),
    }
}

fn first_u64_failure(value: &Map<String, Value>, fields: &[&str]) -> Option<&'static str> {
    fields
        .iter()
        .find_map(|field| canonical_u64_failure(value.get(*field)))
}

fn cursor_failure(value: Option<&Value>) -> Option<&'static str> {
    let Some(cursors) = value.and_then(Value::as_array) else {
        return Some("UNKNOWN_FIELD");
    };
    let mut identities = BTreeSet::new();
    for candidate in cursors {
        let Some(cursor) = object(candidate) else {
            return Some("UNKNOWN_FIELD");
        };
        if let Some(failure) = first_u64_failure(cursor, &["stream_epoch", "acknowledged_position"])
        {
            return Some(failure);
        }
        let identity = (
            cursor.get("stream_id").and_then(Value::as_str),
            cursor.get("stream_epoch").and_then(Value::as_str),
        );
        if !identities.insert(identity) {
            return Some("CURSOR_CONFLICT");
        }
    }
    None
}

fn current_session_failure(
    value: &Map<String, Value>,
    current: Option<CloudLinkSessionContext<'_>>,
) -> Option<&'static str> {
    let current = current?;
    let expected = [
        ("gateway_id", current.gateway_id),
        ("session_id", current.session_id),
        ("session_epoch", current.session_epoch),
        ("credential_generation", current.credential_generation),
    ];
    expected.iter().find_map(|(field, expected)| {
        (value.get(*field).and_then(Value::as_str) != Some(*expected)).then_some("STALE_SESSION")
    })
}

fn canonical_json(value: &Value) -> Option<String> {
    serde_json::to_string(value).ok()
}

fn business_digest(value: &Map<String, Value>) -> Option<String> {
    let projection = json!({
        "message_kind": value.get("message_kind")?,
        "payload": value.get("payload")?,
        "protocol_version": value.get("protocol_version")?,
    });
    let canonical = canonical_json(&projection)?;
    Some(format!(
        "{DIGEST_PREFIX}{:x}",
        Sha256::digest(canonical.as_bytes())
    ))
}

fn valid_digest(value: &str) -> bool {
    value.len() == 71
        && value.starts_with(DIGEST_PREFIX)
        && value[DIGEST_PREFIX.len()..]
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
}

fn delivery_failure(
    value: &Map<String, Value>,
    context: &CloudLinkFixtureContext<'_>,
) -> Option<&'static str> {
    let delivery = object(value.get("delivery")?)?;
    if let Some(failure) = first_u64_failure(delivery, &["stream_epoch", "position"]) {
        return Some(failure);
    }
    let Some(digest) = delivery.get("digest").and_then(Value::as_str) else {
        return Some("INVALID_DIGEST");
    };
    if !valid_digest(digest) {
        return Some("INVALID_DIGEST");
    }
    if business_digest(value).as_deref() != Some(digest) {
        return Some("DIGEST_MISMATCH");
    }
    let prior_json = context.prior_accepted_delivery?;
    let prior: Value = serde_json::from_str(prior_json).ok()?;
    let prior = object(&prior)?;
    let prior_delivery = object(prior.get("delivery")?)?;
    let same_identity = value.get("gateway_id") == prior.get("gateway_id")
        && delivery.get("stream_id") == prior_delivery.get("stream_id")
        && delivery.get("stream_epoch") == prior_delivery.get("stream_epoch")
        && delivery.get("position") == prior_delivery.get("position");
    if same_identity
        && (delivery.get("batch_id") != prior_delivery.get("batch_id")
            || delivery.get("digest") != prior_delivery.get("digest"))
    {
        return Some("DIGEST_CONFLICT");
    }
    None
}

fn valid_signature(value: Option<&Value>) -> bool {
    let Some(signature) = value.and_then(object) else {
        return false;
    };
    !unknown_field(signature, &["key_id", "algorithm", "signature"])
        && signature.get("key_id").and_then(Value::as_str).is_some()
        && signature.get("algorithm").and_then(Value::as_str) == Some("Ed25519")
        && signature
            .get("signature")
            .and_then(Value::as_str)
            .is_some_and(|value| {
                value.len() == 86
                    && value
                        .bytes()
                        .all(|byte| byte.is_ascii_alphanumeric() || byte == b'_' || byte == b'-')
            })
}

fn session_hello_failure(value: &Map<String, Value>) -> Option<&'static str> {
    if unknown_field(
        value,
        &[
            "schema",
            "protocol",
            "message_kind",
            "gateway_id",
            "credential_binding",
            "challenge_id",
            "gateway_key_id",
            "gateway_signature",
            "offered_protocol_versions",
            "client_nonce",
            "resume",
        ],
    ) {
        return Some("UNKNOWN_FIELD");
    }
    if !value
        .get("client_nonce")
        .and_then(Value::as_str)
        .is_some_and(|nonce| {
            nonce.len() == 43
                && nonce
                    .bytes()
                    .all(|byte| byte.is_ascii_alphanumeric() || byte == b'_' || byte == b'-')
        })
    {
        return Some("FIELD_BOUND");
    }
    let binding = object(value.get("credential_binding")?)?;
    if let Some(failure) = canonical_u64_failure(binding.get("generation")) {
        return Some(failure);
    }
    match binding.get("origin_model").and_then(Value::as_str) {
        Some("gateway-signed") => {
            if value.get("gateway_signature").is_none() {
                return Some("AUTHENTICATION_REQUIRED");
            }
            if value
                .get("gateway_key_id")
                .and_then(Value::as_str)
                .is_none()
                || !valid_signature(value.get("gateway_signature"))
            {
                return Some("AUTHENTICATION_INVALID");
            }
        }
        Some("trusted-connector-broker-attestation") => {}
        _ => return Some("AUTHENTICATION_INVALID"),
    }
    cursor_failure(value.get("resume"))
}

fn session_challenge_request_failure(value: &Map<String, Value>) -> Option<&'static str> {
    if unknown_field(
        value,
        &[
            "schema",
            "protocol",
            "message_kind",
            "gateway_id",
            "credential_binding",
            "offered_protocol_versions",
            "client_nonce",
            "resume",
        ],
    ) {
        return Some("UNKNOWN_FIELD");
    }
    if !value
        .get("client_nonce")
        .and_then(Value::as_str)
        .is_some_and(|nonce| {
            nonce.len() == 43
                && nonce
                    .bytes()
                    .all(|byte| byte.is_ascii_alphanumeric() || byte == b'_' || byte == b'-')
        })
    {
        return Some("FIELD_BOUND");
    }
    let binding = object(value.get("credential_binding")?)?;
    if unknown_field(binding, &["credential_id", "generation"]) {
        return Some("UNKNOWN_FIELD");
    }
    if let Some(failure) = canonical_u64_failure(binding.get("generation")) {
        return Some(failure);
    }
    cursor_failure(value.get("resume"))
}

/// Returns whether a string follows `SemVer` 2.0.0 without accepting leading-zero
/// numeric identifiers or empty pre-release/build components.
#[must_use]
pub fn is_strict_semver(value: &str) -> bool {
    let (without_build, build) = value
        .split_once('+')
        .map_or((value, None), |parts| (parts.0, Some(parts.1)));
    if value.matches('+').count() > 1 || build.is_some_and(|part| !valid_semver_ids(part, false)) {
        return false;
    }
    let (core, prerelease) = without_build
        .split_once('-')
        .map_or((without_build, None), |parts| (parts.0, Some(parts.1)));
    if prerelease.is_some_and(|part| !valid_semver_ids(part, true)) {
        return false;
    }
    let components = core.split('.').collect::<Vec<_>>();
    components.len() == 3 && components.into_iter().all(valid_core_number)
}

fn valid_core_number(value: &str) -> bool {
    !value.is_empty()
        && value.bytes().all(|byte| byte.is_ascii_digit())
        && (value == "0" || !value.starts_with('0'))
}

fn valid_semver_ids(value: &str, reject_numeric_leading_zero: bool) -> bool {
    !value.is_empty()
        && value.split('.').all(|identifier| {
            !identifier.is_empty()
                && identifier
                    .bytes()
                    .all(|byte| byte.is_ascii_alphanumeric() || byte == b'-')
                && (!reject_numeric_leading_zero
                    || !identifier.bytes().all(|byte| byte.is_ascii_digit())
                    || identifier == "0"
                    || !identifier.starts_with('0'))
        })
}

fn envelope_failure(
    value: &Map<String, Value>,
    context: &CloudLinkFixtureContext<'_>,
) -> Option<&'static str> {
    if unknown_field(
        value,
        &[
            "schema",
            "protocol",
            "protocol_version",
            "message_kind",
            "gateway_id",
            "session_id",
            "session_epoch",
            "credential_generation",
            "sent_at_ms",
            "expires_at_ms",
            "delivery",
            "message_authentication",
            "traceparent",
            "payload",
        ],
    ) {
        return Some("UNKNOWN_FIELD");
    }
    if value.get("protocol_version").and_then(Value::as_str) != Some("1.0") {
        return Some("UNSUPPORTED_VERSION");
    }
    if let Some(failure) = first_u64_failure(
        value,
        &["session_epoch", "credential_generation", "sent_at_ms"],
    ) {
        return Some(failure);
    }
    if let Some(failure) = current_session_failure(value, context.current_session) {
        return Some(failure);
    }
    if let Some(failure) = delivery_failure(value, context) {
        return Some(failure);
    }
    match value.get("message_kind").and_then(Value::as_str) {
        Some("runtime-manifest-report") => {
            let manifest = object(object(value.get("payload")?)?.get("manifest")?)?;
            if !manifest
                .get("aether_version")
                .and_then(Value::as_str)
                .is_some_and(is_strict_semver)
            {
                return Some("SEMVER_INVALID");
            }
        }
        Some("data-loss") => {
            let payload = object(value.get("payload")?)?;
            let first = payload
                .get("first_lost_position")
                .and_then(Value::as_str)?
                .parse::<u64>()
                .ok()?;
            let last = payload
                .get("last_lost_position")
                .and_then(Value::as_str)?
                .parse::<u64>()
                .ok()?;
            let earliest = payload
                .get("earliest_retained_position")
                .and_then(Value::as_str)?
                .parse::<u64>()
                .ok()?;
            if first > last || last >= earliest {
                return Some("DATA_LOSS_RANGE_INVALID");
            }
        }
        Some("telemetry-batch") => {}
        _ => return Some("UNSUPPORTED_VERSION"),
    }
    None
}

fn non_envelope_failure(
    value: &Map<String, Value>,
    context: &CloudLinkFixtureContext<'_>,
) -> Option<&'static str> {
    match value.get("message_kind").and_then(Value::as_str) {
        Some("session-challenge-request") => {
            return session_challenge_request_failure(value);
        }
        Some("session-hello") => return session_hello_failure(value),
        Some("session-accepted") => return cursor_failure(value.get("resume")),
        Some("session-challenge") => {
            return (!valid_signature(value.get("cloud_signature")))
                .then_some("AUTHENTICATION_INVALID");
        }
        _ => {}
    }
    if value.get("protocol_version").and_then(Value::as_str) != Some("1.0") {
        return Some("UNSUPPORTED_VERSION");
    }
    if let Some(failure) = first_u64_failure(value, &["session_epoch", "credential_generation"]) {
        return Some(failure);
    }
    if let Some(failure) = current_session_failure(value, context.current_session) {
        return Some(failure);
    }
    match value.get("message_kind").and_then(Value::as_str) {
        Some("heartbeat" | "heartbeat-ack") => cursor_failure(value.get("cursors")),
        Some("durable-ack") => {
            if value
                .get("digest")
                .and_then(Value::as_str)
                .is_some_and(valid_digest)
            {
                None
            } else {
                Some("INVALID_DIGEST")
            }
        }
        Some("replay-request") => None,
        _ => Some("UNSUPPORTED_VERSION"),
    }
}

/// Validates one `CloudLink` fixture document and returns the contractual result.
#[must_use]
pub fn validate_cloudlink_fixture<'a>(
    input: &str,
    context: impl CloudLinkContextInput<'a>,
) -> CloudLinkValidationResult {
    let context = context.into_context();
    let Ok(value) = serde_json::from_str::<Value>(input) else {
        return CloudLinkValidationResult {
            accepted: false,
            failure_code: Some("JSON_SYNTAX_ERROR"),
        };
    };
    let Some(value) = object(&value) else {
        return CloudLinkValidationResult {
            accepted: false,
            failure_code: Some("UNKNOWN_FIELD"),
        };
    };
    let is_envelope = matches!(
        value.get("message_kind").and_then(Value::as_str),
        Some("runtime-manifest-report" | "telemetry-batch" | "data-loss")
    );
    let failure = if is_envelope {
        envelope_failure(value, &context)
    } else {
        non_envelope_failure(value, &context)
    };
    CloudLinkValidationResult {
        accepted: failure.is_none(),
        failure_code: failure,
    }
}
